keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,593 @@
+but they will be migrated to use the new fields API.
+
+### Failure and Fallback Strategy
+
+There are different scenarios in which applying an AppArmor profile may fail,
+below are the ones we mapped and their outcome once this KEP is implemented:
+
+| Scenario                                                                                           | API Server Result    | Kubelet Result                                                                                                                                        |
+| -------------------------------------------------------------------------------------------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 1) Using localhost or `runtime/default` profile when container runtime does not support AppArmor.  | Pod created          | The outcome is container runtime dependent. In this scenario containers may 1) fail to start or 2) run normally without having its policies enforced. |
+| 2) Using custom or `runtime/default` profile that restricts actions a container is trying to make. | Pod created          | The outcome is workload and AppArmor dependent. In this scenario containers may 1) fail to start, 2) misbehave or 3) log violations.                  |
+| 3) Using a localhost profile that does not exist on the node.                                      | Pod created          | Container runtime dependent: containers fail to start. Retry respecting RestartPolicy and back-off delay.  Error message in event.                |
+| 4) Using an unsupported runtime profile (i.e. `runtime/default-audit`).                            | Fails validation: pod **not** created. | N/A                                                                                                                                                   |
+| 5) Using localhost profile with invalid (empty) name                                               | Fails validation: pod **not** created. | N/A                                                                                                                                                   |
+| 6) AppArmor is disabled by the host or the build                                                   | Pod created. | Kubelet puts Pod in blocked state.                                                                                                                    |
+
+Scenario 2 is the expected behavior of using AppArmor and it is included here
+for completeness.
+
+Scenario 5 represents the case of failing the existing validation, which is
+defined at [Pod API](#pod-api).
+
+### Version Skew Strategy
+
+All API skew is resolved in the API server. New Kubelets will only use the
+AppArmor values specified in the fields, and ignore the annotations.
+
+#### Pod Creation
+
+If no AppArmor annotations or fields are specified, no action is necessary.
```

Comment:
do we keep the annotations support virtually forever?

Reply:
I like the idea of migrating the annotations and then removing them.


:thought_balloon: This could even cover a Pod that has been happily running since v1.8 or something and the node is still there accumulating uptime. Maybe all of the containers have been restarted over the years and none of the original code is running, but it's still the same Pod and the AppArmor profile still constrains it.

Reply:
Here is the discussion we had for seccomp annotations: https://github.com/kubernetes/kubernetes/pull/109819#discussion_r868722355. I think we should follow the same path here. I'll add a section for it.

> This could even cover a Pod that has been happily running since v1.8 or something and the node is still there accumulating uptime. Maybe all of the containers have been restarted over the years and none of the original code is running, but it's still the same Pod and the AppArmor profile still constrains it.

I thought we don't support in-place kubelet upgrades, so we don't need to worry about this case?